src/lib/work_partitioner/work_partitioner.ml

```diff
@@ -1 +1,244 @@
+open Core_kernel
+
+open struct
+  module Work = Snark_work_lib
+end
+
 module Snark_worker_shared = Snark_worker_shared
+module Zkapp_command_job_pool =
+  Job_pool.Make (Work.Id.Single) (Pending_zkapp_command)
+module Sent_zkapp_job_pool =
+  Job_pool.Make (Work.Id.Sub_zkapp) (Work.Spec.Sub_zkapp.Stable.Latest)
+module Sent_single_job_pool =
+  Job_pool.Make (Work.Id.Single) (Work.Spec.Single.Stable.Latest)
+
+type t =
+  { logger : Logger.t
+  ; transaction_snark : (module Transaction_snark.S)
+  ; single_id_gen : Id_generator.t
+  ; subzkapp_id_gen : Id_generator.t
+  ; pairing_pool : (int64, Combining_result.t) Hashtbl.t
+        (** if one single work from underlying Work_selector is completed but
+           not the other. throw it here. *)
+  ; zkapp_command_jobs : Zkapp_command_job_pool.t
+  ; reassignment_timeout : Time.Span.t
+  ; zkapp_jobs_sent_by_partitioner : Sent_zkapp_job_pool.t
+  ; single_jobs_sent_by_partitioner : Sent_single_job_pool.t
+  ; mutable tmp_slot :
+      (Work.Spec.Single.t * Work.Id.Single.t * Mina_base.Sok_message.t) option
+        (** When receving a `Two works from the underlying Work_selector, store
+            one of them here, so we could schedule them to another worker. *)
+  }
+
+let create ~(reassignment_timeout : Time.Span.t) ~(logger : Logger.t) : t =
+  let module M = Transaction_snark.Make (struct
+    let constraint_constants = Genesis_constants.Compiled.constraint_constants
+
+    let proof_level = Genesis_constants.Compiled.proof_level
+  end) in
+  { logger
+  ; transaction_snark = (module M)
+  ; single_id_gen = Id_generator.create ~logger
+  ; subzkapp_id_gen = Id_generator.create ~logger
+  ; pairing_pool = Hashtbl.create (module Int64)
+  ; zkapp_command_jobs = Zkapp_command_job_pool.create ()
+  ; reassignment_timeout
+  ; zkapp_jobs_sent_by_partitioner = Sent_zkapp_job_pool.create ()
+  ; single_jobs_sent_by_partitioner = Sent_single_job_pool.create ()
+  ; tmp_slot = None
+  }
+
+let epoch_now () = Time.(now () |> to_span_since_epoch)
+
+let reschedule_if_old ~reassignment_timeout
+    (job : _ Work.With_job_meta.Stable.Latest.t) =
+  let scheduled = Time.of_span_since_epoch job.scheduled_since_unix_epoch in
+  let delta = Time.(diff (now ()) scheduled) in
+  if Time.Span.( > ) delta reassignment_timeout then
+    `Stop_reschedule
+      Work.With_job_meta.Stable.Latest.
+        { job with scheduled_since_unix_epoch = epoch_now () }
+  else `Stop_keep
+
+(* NOTE: below are logics for work requesting *)
+let reschedule_old_zkapp_job
+    ~partitioner:
+      ({ reassignment_timeout; zkapp_jobs_sent_by_partitioner; _ } : t) :
+    Work.Spec.Partitioned.Stable.Latest.t Or_error.t option =
+  let%map.Option job =
+    Sent_zkapp_job_pool.remove_until_reschedule
```

Comment:
Check if the pending zkapp pool still contains the parent id for this job (remove it if parent no longer exists because it became irrelevant).

Use some `t.is_relevant` to remove the job if the pending job was rescheduled.
Remove a job from pending zkapp pool in that case too.